src/BLSSignatureChecker.sol

```diff
@@ -72,135 +77,178 @@ contract BLSSignatureChecker is IBLSSignatureChecker {
+            (quorumNumbers.length == params.nonSignerStakeIndices.length),
+            "BLSSignatureChecker.checkSignatures: input quorum length mismatch"
+        );
+
+        require(
+            params.nonSignerPubkeys.length == params.nonSignerQuorumBitmapIndices.length,
+            "BLSSignatureChecker.checkSignatures: input nonsigner length mismatch"
+        );
+
+        require(referenceBlockNumber <= uint32(block.number), "BLSSignatureChecker.checkSignatures: invalid reference block");
+
+        // This method needs to calculate the aggregate pubkey for all signing operators across
+        // all signing quorums. To do that, we can query the aggregate pubkey for each quorum
+        // and subtract out the pubkey for each nonsigning operator registered to that quorum.
+        //
+        // In practice, we do this in reverse - calculating an aggregate pubkey for all nonsigners,
+        // negating that pubkey, then adding the aggregate pubkey for each quorum.
         BN254.G1Point memory apk = BN254.G1Point(0, 0);
-        for (uint i = 0; i < quorumNumbers.length; i++) {
-            if (staleStakesForbidden) {
-                require(
-                    registryCoordinator.quorumUpdateBlockNumber(uint8(quorumNumbers[i])) + delegation.withdrawalDelayBlocks() <= block.number,
-                    "BLSSignatureChecker.checkSignatures: StakeRegistry updates must be within withdrawalDelayBlocks window"
+
+        // For each quorum, we're also going to query the total stake for all registered operators
+        // at the referenceBlockNumber, and derive the stake held by signers by subtracting out
+        // stakes held by nonsigners.
+        QuorumStakeTotals memory stakeTotals;
+        stakeTotals.totalStakeForQuorum = new uint96[](quorumNumbers.length);
+        stakeTotals.signedStakeForQuorum = new uint96[](quorumNumbers.length);
+
+        NonSignerInfo memory nonSigners;
+        nonSigners.quorumBitmaps = new uint256[](params.nonSignerPubkeys.length);
+        nonSigners.pubkeyHashes = new bytes32[](params.nonSignerPubkeys.length);
+
+        {
+            // Get a bitmap of the quorums signing the message, and validate that
+            // quorumNumbers contains only unique, valid quorum numbers
+            uint256 signingQuorumBitmap = BitmapUtils.orderedBytesArrayToBitmap(quorumNumbers, registryCoordinator.quorumCount());
+
+            for (uint256 j = 0; j < params.nonSignerPubkeys.length; j++) {
+                // The nonsigner's pubkey hash doubles as their operatorId
+                // The check below validates that these operatorIds are sorted (and therefore
+                // free of duplicates)
+                nonSigners.pubkeyHashes[j] = params.nonSignerPubkeys[j].hashG1Point();
+                if (j != 0) {
+                    require(
+                        uint256(nonSigners.pubkeyHashes[j]) > uint256(nonSigners.pubkeyHashes[j - 1]),
```

Comment:
@0x0aa0 pls make sure that the offchain code can deal with this

in fact, we should document all interfaces breaking changes since m2 testnet so we know how everything must change in the clients

Reply:
FWIW i did not add this - it was already in this method

Reply:
the DASM on testnet also has this sorting check so I do not think it is breaking